### DISCUSSION TITLE
ML and do - mpc

### DISCUSSION THREAD

**Discussion by ghost** [do-mpc/do-mpc#216] [General]
Hello,

Is it possible to use ml models for estimation / prediction of time series for system modeling with do - mpc?

And if do - mpc doesnt support this yet how we can implement it custom? Meaning use do - mpc and scripts with ml models for the regression problem.

I really admire your work.

Thanks.

**Comment by 4flixt** [comment]
First of all: Thank you for the admiration ❤️ - it is much appreciated!

Regarding your question: Yes it is possible to use ML (e.g. deep learning) models as the system model in do-mpc. We do this often ourselves in our research. 
Unfortunately, it still requires a bit of work on your end currently:

1. Use any ML toolbox to train a model (e.g. Tensorflow with Keras)
2. Retrieve model parameters and structure, for example with a Keras ``model``:
```
model.get_weights()
model.get_config()
```
3. Reimplement the ML model with do-mpc

The last part may seem difficult but it really isn't for many ML models. Remember that a Dense Feedforward neural network is really only a number of matrix vector multiplications. We have also previously implemented GPs or SVM. Most of these approaches have a rather simple computational graph and the challenge lies in training the parameters.

The good news is: Since we use this approach often ourselves, we are aiming to automise it in the future. in particular we want to provide a conversion tool for tensorflow models (or potentially the [ONXX](https://onnx.ai) standard).

Hope this helps!
Best
4flixt

**Reply by ghost** [reply]
Hi 4flixt
Thanks for your answer and comments you provide. I find them really useful and explanatory. Actually, i have already implemented my ml model (RBFN) from scratch with basic numpy. So, I will get back on it !

ps: do - mpc is the best : )
